ray-operator/controllers/ray/expectations/scale_expectation.go

```diff
@@ -0,0 +1,152 @@
+package expectations
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const DefaultHead = ""
+
+// ScaleAction is the action of scale, like create and delete.
+type ScaleAction string
+
+const (
+	// Create action
+	Create ScaleAction = "create"
+	// Delete action
+	Delete ScaleAction = "delete"
+)
+
+const (
+	// GroupIndex indexes pods within the specified group in RayCluster
+	GroupIndex = "group"
+	// RayClusterIndex indexes pods within the RayCluster
+	RayClusterIndex = "raycluster"
+)
+
+// RayClusterScaleExpectation is an interface that to set and wait on expectations of RayCluster groups scale.
+type RayClusterScaleExpectation interface {
+	ExpectScalePod(rayClusterName, group, namespace, name string, action ScaleAction)
+	IsSatisfied(ctx context.Context, rayClusterName, group, namespace string) bool
+	Delete(rayClusterName, namespace string)
+}
+
+func newRayPodIndexer() cache.Indexer {
+	return cache.NewIndexer(rayPodKey, cache.Indexers{GroupIndex: groupIndexFunc, RayClusterIndex: rayClusterIndexFunc})
+}
+
+func NewRayClusterScaleExpectation(client client.Client) RayClusterScaleExpectation {
+	return &realRayClusterScaleExpectation{
+		Client:     client,
+		itemsCache: newRayPodIndexer(),
+	}
+}
+
+type realRayClusterScaleExpectation struct {
+	client.Client
+	itemsCache cache.Indexer
+}
+
+func (r *realRayClusterScaleExpectation) ExpectScalePod(rayClusterName, group, namespace, name string, action ScaleAction) {
+	_ = r.itemsCache.Add(&rayPod{
```

Comment:
The source of errors in `Delete`/`Create` originates from the `keyFunc`. We strictly limit the data type stored in the itemsCache to be rayPod. Therefore, the types of the objects for rayPodKey, groupIndexFunc, and rayClusterIndexFunc must be rayPod. This ensures that we won't encounter errors throughout the entire process. By doing this, we can avoid issues right from the beginning.

Additionally, we should not try to avoid panic, as a panic indicates that there is a problem in our code: specifically, that we have not ensured the type of data stored in the itemsCache is rayPod.

Reply:
The source of errors in `Delete()`/`Create()` originates from the `keyFunc`. 
https://github.com/kubernetes/client-go/blob/master/tools/cache/store.go#L170

We strictly limit the data type stored in the `itemsCache` to be `rayPod`. Therefore, the types of the objects for `rayPodKey`, `groupIndexFunc`, and `rayClusterIndexFunc` must be `rayPod`. This ensures that we won't encounter errors throughout the entire process. By doing this, we can avoid issues right from the beginning.

Additionally, we should not try to avoid panic, as a panic indicates that there is a problem in our code: specifically, that we have not ensured the type of item stored in the `itemsCache` is `rayPod`.

Reply:
You're right, but I think this requires a lot of context and assumption. Maybe we can change the `ExpectScalePod` to return an error, just like how kubernetes does.

Reply:
I initially thought this wouldn’t count as a fatal error; if there’s an issue with the expectation, we could still fall back to the original method. However, on second thought, that might make things a bit complicated, so I now think treating it as a fatal error and letting it panic is also fine.